### PR TITLE
AddKoreAttribute as a Frontend Kompiler Pass

### PR DESCRIPTION
**Pull request by Robertorosmaninho**
Part of #3444

This PR creates a new compiler pass that adds the Kore attributes to productions before `ModuleToKore`.  This is part of the effort to refactor the `ModuleToKore` infrastructure to simplify it by decoupling functions and actions that can be done before the translation and that aren't related to (kore) code emission.

Besides extracting the `AddKoreAttrubutes` function from `ModuleToKore.java`, this PR also simplifies function calls such as `isFunctional` and `isConstructor` only to test if the attribute is present on the given production.

### CODE REVIEW

File: kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java
Comment (Robertorosmaninho) [inline kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java]: For some reason, this Exception is thrown due to finding 0 production for KLabel.

File: kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java
Comment (Robertorosmaninho) [inline kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java]: I'm having issues here, probably due to the Module transformer, as even though I can add the Attributes to the production, they don't appear updated here.

**Comment by Robertorosmaninho** [comment]
I'm currently blocked on the two pending issues mentioned above.
I'm testing it with this [test](https://github.com/runtimeverification/k/blob/master/k-distribution/tests/regression-new/cell-sort-haskell/test.k) as it was one of the first from `regression-new` to fail within this implementation.

The `definition.kore` reflect the second issue, as the symbols don't contain the appropriate attributes inserted by the AddKoreAttributes pass.

File: kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java]: You want to call `mutable` from org.kframework.Collections which we import statically usually.

File: kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java]: Why is this being put here? This should be a separate refactoring. If you are somehow depending on these axioms being present, they should be placed in a separate pass in a separate pr that is refactored first and run first.

File: kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java
Comment (dwightguth) [reply kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java]: I can't tell you why that might be without knowing what klabel it's throwing the error for. In general it's always better to share the entire exact error message rather than to just summarize in a way that leaves information out.

**Comment by dwightguth** [comment]
As far as the issue with the module transformer is concerned, it is a known issue that passes that only edit the attributes of sentences in a module get cached incorrectly by the code that helps cache module transformations, as a result of which the changes to the attribute get discarded.

You have two options:

1. In a separate PR, try to fix the issue with module attribute caching. I don't remember exactly why I had trouble doing this when I tried, but I remember it wasn't as easy as I hoped it would be.
2. Bundle this pass together in the same module transformer as another change that edits something other than the attributes of every module it edits attributes in. This will bypass the issue with caching and correctly give back the right info.

**Comment by Robertorosmaninho** [comment]
> As far as the issue with the module transformer is concerned, it is a known issue that passes that only edit the attributes of sentences in a module get cached incorrectly by the code that helps cache module transformations, as a result of which the changes to the attribute get discarded.
> 
> You have two options:
> 
> 1. In a separate PR, try to fix the issue with module attribute caching. I don't remember exactly why I had trouble doing this when I tried, but I remember it wasn't as easy as I hoped it would be.

That's interesting, could you provide me with more details about it?
- When/how did you face this issue before?
- What have you tried?
- How do we cache these module transformations?

**Comment by Robertorosmaninho** [comment]
Blocked on #3523.
We need to fix the Transformation caching to update the modules when we only modify the sentence attributes!

File: kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java
Comment (Robertorosmaninho) [reply kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java]: Addressing it in #3529

File: kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java
Comment (Robertorosmaninho) [inline kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java]: This will be optimized in a future PR. The idea is to move this and the production's functions to the module definition and call it here and in the ModuleKore file

**Review by radumereuta — APPROVED** [review]
lgtm, but I would like it if someone else could have a look.

File: kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java]: This is now dead.

File: kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java]: this is now dead.

File: kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java]: this is now dead.

File: kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java]: let's share this rather than duplicating it.

File: kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java]: same here: don't duplicate

**Review by dwightguth — CHANGES_REQUESTED** [review]

File: kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java
Comment (Robertorosmaninho) [reply kernel/src/main/java/org/kframework/compile/AddKoreAttributes.java]: Addressed in https://github.com/runtimeverification/k/pull/3522/commits/4e062801eea8b1d6fd165783f4fa7cb7bbc6b208

**Review by dwightguth — APPROVED** [review]

**Pull request merged**
